Instruction Selection, Object Files, and More

LensPlaysGames opened the pull request:
First off, I'll just start by saying I don't even know if this will end up getting merged. The code is quite messy. If I *can* manage to clean it up, however, that will be rather cool. That means the compiler will be able to emit object files which a linker can then use to generate an executable. This is basically "one layer past" assembly, which we currently emit. This PR aims to provide support for an object file backend for x86_64 while retaining the assembly backend.

LensPlaysGames commented:
Progress Report: `simple_math.int` is assembled completely into final machine code; it could be embedded in an executable file as is. Assembling `simple_pointer.int` succeeds, and all instructions are represented in the final machine code, however there are multiple relocations not applied, leaving the final code incorrect. `greatest_common_denominator.int` also succeeds with similar relocation issues.

TL;DR: We are assembling x86_64 instructions and operands into proper machine code, but that's kind of it. We are missing large portions of required functionality, like keeping track of relocations, symbol entries, etc.

Going forward:
1. Altering the machine-code-producing `mcode_*` functions to append to some sort of byte buffer instead of writing to a file would be ideal
2. Create some sort of meta, in-memory object file structure that keeps track of things like defined symbols, relocations, sections, etc (like our very own lil' BFD hehe)

LensPlaysGames commented on src/codegen/x86_64/arch_x86_64.c:
```diff
@@ -380,8 +383,1257 @@ static enum IndirectJumpType negate_jump(enum IndirectJumpType j) {
+      fwrite(&imm8, 1, 1, context->machine_code);
+    } break;
+    case r16: {
+      // 0x81 /5 iw
```
Missing `0x66 + ` (in the comment)

LensPlaysGames replied:
Fixed as of [1234801](https://github.com/LensPlaysGames/Intercept/pull/66/commits/1234801a026b424dbe55d3b1541a35e70c6c88f0)

LensPlaysGames commented on src/codegen/x86_64/arch_x86_64.c:
```diff
@@ -380,8 +383,1257 @@ static enum IndirectJumpType negate_jump(enum IndirectJumpType j) {
+      // the bit extension.
+      uint8_t address_regbits = regbits(address_register);
+      uint8_t destination_regbits = regbits(destination_register);
+      uint8_t rex = rex_byte(true, REGBITS_TOP(address_regbits), false, REGBITS_TOP(destination_regbits));
```
This looks like `r64` handling, but it's under `r8`. Maybe copy-and-paste error? I don't think there should *always* be a REX prefix for `r8/m8 to r8` opcode...

LensPlaysGames commented on src/codegen/x86_64/arch_x86_64.c:
```diff
@@ -380,8 +383,1257 @@ static enum IndirectJumpType negate_jump(enum IndirectJumpType j) {
+      // REX.W + 0x8b /r
+      uint8_t op = 0x8b;
+
+      // Encode a REX prefix if either of the ModRM register descriptors need
+      // the bit extension.
```
Incorrect comment.

LensPlaysGames replied:
[a91f737](https://github.com/LensPlaysGames/Intercept/pull/66/commits/a91f7379161d408eb0c376604444410ea65302c1)

LensPlaysGames commented on src/codegen/x86_64/arch_x86_64.c:
```diff
@@ -380,8 +383,1257 @@ static enum IndirectJumpType negate_jump(enum IndirectJumpType j) {
+        // one byte using a Mod value of 0b01, allowing for only one byte
+        // to be written for the displacement.
+
+        // Mod == 0b10  ->  register + disp8
```
Incorrect comment; should be `0b01` (See Section 2.1.5 in Chapter 2 of Vol. 2A of the Intel x86_64 SDM Manual Combined Volumes)

![image](https://user-images.githubusercontent.com/69637718/235278031-ac503841-93c8-4dae-8ee5-bc19dba43636.png)

LensPlaysGames replied:
[a91f737](https://github.com/LensPlaysGames/Intercept/pull/66/commits/a91f7379161d408eb0c376604444410ea65302c1)

LensPlaysGames commented on src/codegen/x86_64/arch_x86_64.c:
```diff
@@ -380,8 +383,1257 @@ static enum IndirectJumpType negate_jump(enum IndirectJumpType j) {
+    // PUSH imm8:  0x6a ib
+    // PUSH imm16: 0x68 iw
+    // PUSH imm32: 0x68 id
+    // PROBABLY FALSE LENS_R THOUGHT: I think imm16 uses 0x66 prefix
```
This is no longer probably false and is now I think a fact.

> REX prefixes consist of 4-bit fields that form 16 different values. The W-bit field in the REX prefixes is referred to as
REX.W. If the REX.W field is properly set, the prefix specifies an operand size override to 64 bits. **Note that software
can still use the operand-size 66H prefix to toggle to a 16-bit operand size.** However, setting REX.W takes precedence over the operand-size prefix (66H) when both are used.

See Section 3.6.1 Operand Size and Address Size in 64-Bit Mode in Vol. 1 Chapter 3 of the Intel x86_64 SDM

LensPlaysGames replied:
Resolved in [12e5e48](https://github.com/LensPlaysGames/Intercept/pull/66/commits/12e5e484bc72f83fdb24d77858e49f4daa18be4d)

LensPlaysGames commented on src/codegen/x86_64/arch_x86_64.c:
```diff
@@ -651,6 +1949,8 @@ static void femit
  enum Instruction instruction,
  ...)
 {
+  mcode_none(context, instruction);
```
This isn't within an `#ifdef`

LensPlaysGames replied:
Fixed as of [2d172ce](https://github.com/LensPlaysGames/Intercept/pull/66/commits/2d172cea524c3697386250923acb811c58f80055)

LensPlaysGames committed:
[x86_64] SIB byte helper

LensPlaysGames committed:
[Sema] Prevent NULL dereference

LensPlaysGames committed:
[x86_64/MSWIN] Bugfixes that allow for more stack-based arguments

LensPlaysGames committed:
[x86_64/GObj] More machine code encodings

`sub` in `imm_to_mem` form.
`mov` in `reg_to_name` form.

LensPlaysGames committed:
[GObj/COFF] Actually generate string table instead of truncating

LensPlaysGames committed:
[Minor] Remove debug print statements

LensPlaysGames committed:
[GObj] `.rodata` handling for COFF and ELF

LensPlaysGames committed:
[x86_64/GObj] Create and use `.rodata` section for string literals

LensPlaysGames committed:
[x86_64] Abstract some architecture-specific stuff away from assembly emission

LensPlaysGames committed:
[x86_64] ¡No more varargs!

LensPlaysGames committed:
[x86_64] Don't allocate and free register_pool since we aren't using it

LensPlaysGames committed:
[x86_64] Rename `femit` to `femit_none`

LensPlaysGames committed:
[Codegen] A (very) basic machine IR

This is definitely "ill-formed" and will require lots of cleanup, but
it is a step in the right direction. "Instruction selection" now
becomes it's own pass which we can easily expand upon and make
better. We are still doing RA on the regular IR, which is a bit
scuffed, so we'll have to change that at some point. Overall, this
helps to share more code between each target that supports the x86_64
architecture.

Again, this is just a very basic, beginner-level MIR that we will
make better over time.

LensPlaysGames commented:
Here comes the boom

LensPlaysGames committed:
[Codegen] Expansion upon that basic MIR, among other things

Also separate mcode and femit code entirely in the x86_64 backend;
that is, the GObj backend and GAS backend are getting closer and
closer to actually being separate backends, hehe.

LensPlaysGames committed:
[MIR/Minor] Mildly better debug printout

LensPlaysGames committed:
[MIR] Name operand type, among other things

LensPlaysGames committed:
[MIR] WIP `select_instructions2`, an actual MIR->MIR<arch> pass

Basically, we want to have MIR able to represent generic machine
instructions as well as specific machine instructions, and we want
each architecture/ISA to set up patterns to convert one-or-more of
these generic MIR instructions into one-or-more specific MIR
instructions. `select_instructions2` is basically those patterns but
in code. The idea is that A. RA will move until after all this
happens, and operate on MIR instead of IR and B. each target
architecture can more easily share arch-specific code, RA, etc, and
then each target supported for that arch (i.e. COFF, ELF, GAS, etc) is
just a simple translation layer from these MIR<arch> instructions.

Sirraide commented on src/codegen.c:
```diff
@@ -809,6 +809,7 @@ bool codegen
   if (!code) ICE("codegen(): failed to open file at path: \"%s\"\n", outfile);
 
   CodegenContext *context = codegen_context_create(ast, format, call_convention, dialect, code);
+
```
Big change. Approved.

Sirraide commented on src/codegen/codegen_forward.h:
```diff
@@ -60,4 +60,72 @@ enum ComparisonType {
 };
 
+/// All instructions that take two arguments.
+#define ALL_BINARY_INSTRUCTION_TYPES(F) \
```
I would have left these tables in their respective headers (intermediate_representation.h and machine_ir.h), but I suppose they could also go here. codegen_forward.h was mostly intended for forward declarations to break header dependency cycles

LensPlaysGames replied:
That's also why I needed `machine_ir_forward.h`, but I never thought about just putting those in `codegen_forward.h` lol

LensPlaysGames replied:
I don't know how I thought that was a response to the comment above lol but that was meant for the one down below regarding the header not being needed. 

If you want to try and put the IR macros back in their header, then by all means go ahead, but the current situation took enough fighting with the stupid preprocessor and circular includes that I've given up trying to get that to work. I hate includes so much when projects start getting bigger like this.

Sirraide commented on src/codegen/coff.h:
```diff
@@ -0,0 +1,497 @@
+  // NOTE: Also known as "characteristics" field in some docs.
+  // See HDR_* defines above
+  uint16_t f_flags;
+} __attribute__((packed)) coff_header;
```
This won’t work with MSVC; it uses a pragma for that instead (`#pragma pack(1)` right before the struct decl should do it I think). To fix this, I would recommend something like this:
```c
/// In utils.h:
#if !defined(_MSC_VER) || defined(__clang__)
#    define PACKED __attribute__((packed))
#else
#    define PACKED _Pragma("pack(1)")
#endif

/// In the struct decl:
typedef struct PACKED coff_header { 
    ... 
};
```

The `PACKED` macro must be placed in exactly that position—between the `struct` keyword and the struct name, or else it won’t work.

Note also that `clang-cl` likes to pretend to be MSVC, so we should make sure to handle that properly. I’m not sure the pragma version works w/ `clang-cl`, but at the very least, it doesn’t work with regular Clang.

Sirraide replied:
That or we could just... drop MSVC support so I can start using `typeof` in vector.h. Just saying.

LensPlaysGames replied:
Yeah I stole the actual struct definitions from an old test project; I didn't think about the packed attribute. Tysm for the pragma example, I have no idea how to use them really, hehe. I'll be sure to fix this ASAP

LensPlaysGames replied:
See [def033c](https://github.com/LensPlaysGames/Intercept/pull/66/commits/def033c39dea0bf1df17d8daf34589cd5d4bd39e)

Sirraide commented on src/codegen/coff.h:
```diff
@@ -0,0 +1,497 @@
+  uint32_t entry;
+  uint32_t text_start;
+  uint32_t data_start;
+} __attribute__((packed)) coff_opt_header;
```
Same problem here

Sirraide commented on src/codegen/coff.h:
```diff
@@ -0,0 +1,497 @@
+  /// Number of Line Number table entries
+  uint16_t s_nlnno;
+  int32_t s_flags;
+} __attribute__((packed)) coff_section_header;
```
And here

Sirraide commented on src/codegen/coff.h:
```diff
@@ -0,0 +1,497 @@
+  uint32_t r_symndx;
+  /// Type of relocation
+  uint16_t r_type;
+} __attribute__((packed)) coff_relocation_entry;
```
At this point you know what to do

Sirraide commented on src/codegen/coff.h:
```diff
@@ -0,0 +1,497 @@
+  } l_addr;
+  /// Line number
+  uint16_t l_lnno;
+} __attribute__((packed)) coff_line_number_entry;
```
Here as well

Sirraide commented on src/codegen/coff.h:
```diff
@@ -0,0 +1,497 @@
+
+/**
+ * The Section table name field and the Symbol table name field are
+ * actaully more complicated than was detailed above, they in fact look
```
actaully

Sirraide commented on src/codegen/coff.h:
```diff
@@ -0,0 +1,497 @@
+  char n_sclass;
+  /// Auxiliary Count
+  char n_numaux;
+} __attribute__((packed)) coff_symbol_entry;
```
More packed structs

Sirraide commented on src/codegen/coff.h:
```diff
@@ -0,0 +1,497 @@
+  uint8_t  selection;
+  uint8_t  unused;
+  uint16_t number_high;
+} __attribute__((packed)) coff_aux_section;
```
Still more

Sirraide commented on src/codegen/coff.h:
```diff
@@ -0,0 +1,497 @@
+      uint32_t offset;
+    };
+  };
+} __attribute__((packed)) coff_string_entry;
```
At this point, I’m just gonna mark these 3 in one fell swoop

Sirraide commented on src/codegen/generic_object.c:
```diff
@@ -0,0 +1,627 @@
+  return object ? object->sections.data : NULL;
+}
+
+Section *get_section_by_name(const Sections sections, const char *name) {
```
I’m just gonna comment on this here, otherwise we’ll probably end up w/ 9000 comments about this, but I feel like we should really be using `span`/`string` instead of C-style strings. That would also help get rid of memory leaks because we’d know to always free a `string` if it goes out of scope and isn’t referenced elsewhere etc...

Sirraide commented on src/codegen/generic_object.c:
```diff
@@ -0,0 +1,627 @@
+  return NULL;
+}
+
+/// Write 1 byte of data to code.
+void sec_write_1(Section *section, uint8_t value) {
+  vector_push(section->data.bytes, value);
+}
+/// Write 2 bytes of data to code.
+void sec_write_2(Section *section, uint8_t value0, uint8_t value1) {
+  vector_push(section->data.bytes, value0);
+  vector_push(section->data.bytes, value1);
+}
+/// Write 3 bytes of data to code.
+void sec_write_3(Section *section, uint8_t value0, uint8_t value1, uint8_t value2) {
+  vector_push(section->data.bytes, value0);
+  vector_push(section->data.bytes, value1);
+  vector_push(section->data.bytes, value2);
+}
+/// Write 4 bytes of data to code.
+void sec_write_4(Section *section, uint8_t value0, uint8_t value1, uint8_t value2, uint8_t value3) {
+  vector_push(section->data.bytes, value0);
+  vector_push(section->data.bytes, value1);
+  vector_push(section->data.bytes, value2);
+  vector_push(section->data.bytes, value3);
+}
```
I’m sure we could do some horrible stuff w/ `_Generic` to overload this if you’re interested :eyes:

Sirraide commented on src/codegen/generic_object.c:
```diff
@@ -0,0 +1,627 @@
+}
+/// Write n bytes of data from buffer to code.
+void sec_write_n(Section *section, void* buffer_, size_t n) {
+  // TODO: Test if this manual unrolling is actually any faster.
```
If the loop has a size of 4 that’s known at compile time, I’m 99% sure the compiler’s gonna unroll it anyway

Sirraide commented on src/codegen/generic_object.c:
```diff
@@ -0,0 +1,627 @@
+  }
+
+  // Build elf64_rela relocations
+  Vector(elf64_rela) relocations = {0};
```
I don’t think this is ever deleted

LensPlaysGames replied:
It wasn't :P

See [7ef73fd](https://github.com/LensPlaysGames/Intercept/pull/66/commits/7ef73fd5eca55d01c981c675bccab58abf78041b)

Sirraide commented on src/codegen/generic_object.c:
```diff
@@ -0,0 +1,627 @@
+  size_t data_offset = hdr.e_shoff + (sizeof(elf64_shdr) * hdr.e_shnum);
+
+  // Collect symbols.
+  Vector(elf64_sym) syms = {0};
```
I’m pretty sure this vector is just leaked.

LensPlaysGames replied:
👀
https://github.com/LensPlaysGames/Intercept/pull/66/files/560e1a2b3f2c73c2cd875b0a6e98a5fecd1c6b1b#diff-c8f3c875a9398b7ac77ec2da5b3724b7e6c24b1defd0ddccf2978d6cd6ae8eb2R350-R352

Sirraide commented on src/codegen/intermediate_representation.h:
```diff
@@ -141,7 +74,9 @@ void mark_used(IRInstruction *usee, IRInstruction *user);
   enum IRType kind;
+
   Register result;
+  MIRInstruction *machine_inst;
```
See, this is exactly what the codegen_forward.h header is for. A forward declaration there would suffice, and then this header wouldn’t have to include any MIR headers for this

Sirraide commented on src/codegen/intermediate_representation.h:
```diff
@@ -3,6 +3,7 @@
 
 #include <codegen.h>
 #include <codegen/codegen_forward.h>
+#include <codegen/machine_ir_forward.h>
```
This include can be avoided (see next comment)

Sirraide commented on src/codegen/machine_ir.c:
```diff
@@ -0,0 +1,326 @@
+#include <codegen/machine_ir.h>
+
+#include <codegen.h>
+#include <codegen/intermediate_representation.h>
```
Half of these use "", the other half <>

LensPlaysGames replied:
Oops! Automatic includes are nice, but I do wish I could tell it to use brackets and not quotes haha. Since been removed

Sirraide commented on src/codegen/machine_ir.c:
```diff
@@ -0,0 +1,326 @@
+
+MIROperand mir_op_reference(MIRInstruction *m_inst, IRInstruction *inst) {
+  ASSERT(inst, "Invalid argument");
+  if (!inst->machine_inst) {
+    ir_femit_instruction(stdout, inst);
+    ICE("Must translate IRInstruction into MIR before taking reference to it.");
```
Does this play nicely w/ forward references (e.g. branches and PHIs containing forward references due to e.g. loop codegen)?

LensPlaysGames replied:
Almost certainly not. What is the solution to that? Do we need like "relocations" for the MIR, or is there some easier solution?

Sirraide replied:
VRegs or some other index-based solution is the only thing that I can think of unfortunately... That or keep track of a list of instructions that need to be fixed up later, but I wouldn’t really call that a ‘solution’, candidly.

Sirraide commented on src/codegen/machine_ir.c:
```diff
@@ -0,0 +1,326 @@
+MIRInstruction *mir_makenew(uint32_t opcode) {
+  MIRInstruction *mir = calloc(1, sizeof(*mir));
+  ASSERT(mir, "Memory allocation failure");
+  mir->id = ++mir_alloc_id;
```
This `mir_alloc_id` should probably be part of the codegen context

LensPlaysGames replied:
I'm not really arguing it's a bad idea to get rid of the statics here---that's definitely a good idea---but I do wonder what it would accomplish

Sirraide replied:
> I'm not really arguing it's a bad idea to get rid of the statics here---that's definitely a good idea---but I do wonder what it would accomplish

Nothing unless you’re planning to be able to codegen several programs/files concurrently. If you care about multithreading at all, then these should all go in the context because that’s what it’s for, otherwise none of that matters to begin with and everything that’s in the context could just as well be a bunch of globals. My point here is, it doesn’t make that much sense to put half of this stuff in the context and make the other half global variables. It’s either one way or the other, but mixing both unfortunately doesn’t end up accomplishing much of anything because you get both the inconvenience of having to pass a context to every function and yet no thread safety in return...

Sirraide commented on src/codegen/machine_ir.h:
```diff
@@ -0,0 +1,93 @@
+
+typedef struct MIRValue_x86_64 {
+  uint8_t instruction_form;
+  uint16_t instruction;
```
Swap the order of these two to reduce the size of this struct by 1 byte of padding

Sirraide commented on src/codegen/machine_ir_forward.h:
```diff
@@ -0,0 +1,44 @@
+#ifndef MACHINE_IR_FORWARD_H
+#define MACHINE_IR_FORWARD_H
```
I feel like this header doesn’t really need to exist. All forward declarations should just go in regular codegen_forward.h and the rest in machine_ir.h. Unless there is a specific reason to split these up into two, of course

Sirraide reviewed (commented):
This is only a partial review. I’ll finish this later probably. Left off at machine_ir_forward.h.

LensPlaysGames committed:
[Minor] Formatting

LensPlaysGames committed:
[x86_64] Fix encoding of `imm-to-reg` form of `mov`

LensPlaysGames committed:
[MIR] Lots of changes

We are slowly working towards having an MIR that fully represents the
program while at the same time being closer in representation to the
underlying ISA that we will be emitting code for. From here, there's
lots of complex stuff to do, like moving register allocation to MIR
instead of IR. And also pattern matching for instruction selection
using a DSL or something would be glorious.

LensPlaysGames committed:
Remove `machine_ir_forward.h` header

LensPlaysGames committed:
[MSVC] PACKED macro

LensPlaysGames committed:
[GObj] Add missing `vector_delete` for ELF translation

LensPlaysGames committed:
[MIR] `makecopy` no longer takes opcode param

LensPlaysGames committed:
[Build] Make incompatible pointer types warnings an error

LensPlaysGames committed:
[Utils] `DBGASSERT`

LensPlaysGames committed:
[MIR] Function and block abstraction

LensPlaysGames committed:
[Minor] Separate printing of MIR operand from printing of MIR instruction

LensPlaysGames committed:
[MIR/x86_64] Slightly fixup call generation

It's still a mess, but at least it's getting better.

LensPlaysGames committed:
[MIR] Add missing `stddef.h` include for `size_t`

Honestly I probably should just use `usz` included from utils.h

LensPlaysGames committed:
[MIR] Print hardware vs virtual register consistently

Sirraide commented on src/codegen/machine_ir.h:
```diff
@@ -0,0 +1,180 @@
+  /// new MIR with) the arch-specific opcodes.
+  uint32_t opcode;
+
+  uint8_t operand_count;
+  union {
+    MIROperand arr[MIR_OPERAND_SSO_THRESHOLD];
+    Vector(MIROperand) vec;
+  } operands;
```
In my version of this, I opted to mark operands as unused using a ‘null operand’ rather than having a separate operand count. I just feel that having an operand count might lead to confusion if the long representation is used because then we have both the operand count and the size of the vector. Moreover, I haven’t checked this, but I’m almost certain that this operand count ends up adding 4 bytes to the struct (because of padding).

LensPlaysGames replied:
I'm confused; then how would you know when something is in the vector vs in the array? Are you saying it shouldn't be a union?

Sirraide replied:
In my version of this, an operand is 16 bytes, so I’m reusing the `kind` member of the first operand as a union tag (and the vector is spread across the three operands, but you can also just move it to offset 8 into the union or whatever)

Sirraide replied:
If an operand is only 8 bytes, then this doesn’t work ofc, but iirc operands are 16 bytes here

Sirraide commented on src/codegen/machine_ir.h:
```diff
@@ -0,0 +1,180 @@
+typedef struct MIRBlock {
+  string name;
+
+  // Index within `instructions` of MIRFunction pointing to first
+  // instruction executed in basic block.
+  MIRRegister entry;
+  // Index within `instructions` of MIRFunction pointing to last
+  // instruction of basic block.
+  MIRRegister exit;
```
I can see this becoming a problem if one MIR instruction is replaced with several instructions or if multiple instructions are folded into one, as is a common operation during instruction selection.

LensPlaysGames replied:
Interesting; in my head I was thinking that by the time we get to MIR we wouldn't really be doing much replacing of instructions or anything, as that would all happen in the IR or during the pattern matching and creation of the lowered MIR from the general MIR. If we do end up having to do a lot of the same operations on the MIR as on the IR (like replacing instructions, keeping track of uses, etc), then yeah this isn't a good move. I just figured MIR was pretty much static as everything happened in IR (like lowering) or in genMIR->loweredMIR stage.

Sirraide replied:
Well, you still need to do instruction selection on the MIR. I’d say we either do blocks just like in the IR or abolish blocks completely, but I’m not sure this approach here would work too well...

LensPlaysGames replied:
The instruction selection happens during the genMIR -> loweredMIR stage; this stage will eventually be declarative through the pattern table, where patterns of generic MIR instructions will be used to add new loweredMIR instructions to a new set of loweredMIR. I see no reason to mutate the generic MIR into lowered MIR when we can use it as an immutable input to possibly generate multiple sets of lowered MIR, if a user may want to compile for multiple arch/target combos

LensPlaysGames replied:
Okay I redact my statement we should absolutely mutate the MIR because of virtual register numbering and reference invalidation LMAO

Sirraide commented on src/codegen/machine_ir.h:
```diff
@@ -0,0 +1,180 @@
+  string name;
+
+  /// Flat list of all instructions in function.
+  MIRInstructionVector instructions;
```
This being a vector rather than a list definitely makes sense, and it might well make sense to perform instruction selection into a completely new vector because we only ever have to modify the last couple instructions during ISel. We may have to split this into several vectors though, one for each block (see also the previous comment in this file)

LensPlaysGames replied:
How then would one find an instruction by virtual register?

Right now we use the virtual register of an instruction as an index into this vector (and we use the size of this vector to set the virtual register when pushing a new instruction). This means that all the virtual register references have constant lookup and (eventual) MIR register allocation would have a much quicker time resolving all these references. Why do the instructions need to be stored in separate vectors?

Sirraide replied:
> How then would one find an instruction by virtual register?

Not at all. Assuming the RA works properly, there is no need for that since instructions communicate over what registers they use and place their results in

Sirraide replied:
> Why do the instructions need to be stored in separate vectors?

Mostly for the reasons discussed in the previous comment in this file (if we change how blocks are handled)

LensPlaysGames replied:
> > How then would one find an instruction by virtual register?
> 
> Not at all. Assuming the RA works properly, there is no need for that since instructions communicate over what registers they use and place their results in

Interesting. So you're saying RA would keep a "translation list" of vreg->hwreg and then any time it encounters that vitrual register it just looks it up in that list and replaces it with the corresponding hardware register? And in this way a virtual register operand would be resolved without having to reference the machine instruction that populates that register at all

Sirraide replied:
The RA’s job is to assign hw regs to the vregs, but yes

Sirraide commented on src/codegen/machine_ir.h:
```diff
@@ -0,0 +1,180 @@
+#define MIR_FOREACH_INST_IN_BLOCK(block, it_name)                       \
+  const size_t block##idx_start = (size_t)(block)->entry - (size_t)MIR_ARCH_START; \
+  const size_t block##idx_end = (size_t)(block)->exit - (size_t)MIR_ARCH_START; \
+  DBGASSERT(block##idx_start < block->function->instructions.size, "MIRBlock entry (%Z) is not less than size of machine instruction vector (%Z)", block##idx_start, block->function->instructions.size); \
+  DBGASSERT(block##idx_end < block->function->instructions.size, "MIRBlock exit (%Z) larger than size of machine instruction vector (%Z) of function %S", block##idx_end, block->function->instructions.size, block->function->name); \
+  DBGASSERT(block##idx_start <= block##idx_end, "MIRBlock exit is before it's entry: %Z < %Z", block##idx_end, block##idx_start); \
+  MIRInstruction *it_name = block->function->instructions.data[block##idx_start]; \
+  for (size_t it = block##idx_start; it <= block##idx_end; ++it, it_name = block->function->instructions.data[it])
+
```
This is *probably* fine, but I’d generally always use `CAT` rather than `##` directly, just in case `block` or `it_name` ends up being a macro.

Sirraide replied:
Also, the possessive form of ‘it’ is ‘its’, and not ‘it’s’ ;Þ

LensPlaysGames replied:
Using `CAT(block, idx_start)` and `CAT(block, idx_end)` in place of every `##` didn't work; maybe I'm using it wrong or misremembering

Sirraide replied:
That’s strange. It should definitely work...

Sirraide reviewed (commented):
Still not done, but submitting these for now

Sirraide commented on src/error.h:
```diff
@@ -43,6 +43,13 @@ typedef struct {
+# define DBGASSERT(...)
+#else
+# define DBGASSERT(cond, ...) \
+  ((cond) ? (void)(0) : (raise_fatal_error_impl(__FILE__, PRETTY_FUNCTION, __LINE__, false, false, #cond, "" __VA_ARGS__), BUILTIN_UNREACHABLE()))
```
Is there any reason why this isn’t just defined like so in the `#else` part?
```c
#    define DBGASSERT(...) ASSERT(__VA_ARGS__)
```

LensPlaysGames replied:
Nope; that's probably a good idea haha

Sirraide commented on src/codegen/x86_64/arch_x86_64_isel.h:
```diff
@@ -0,0 +1,13 @@
+#include <codegen/codegen_forward.h>
+#include <codegen/machine_ir.h>
+
+/// Given Intercept IR, return a vector of machine instructions for x86_64 that represents the program.
+MIRInstructionVector select_instructions(CodegenContext *context);
+
+/// Given freshly-lowered MIR, return a vector of machine instructions for x86_64 that represents the program.
+MIRFunctionVector select_instructions2(MIRFunctionVector input);
```
Er, why do two of these exist? Generally, it should probably always go   
AST -> IR (-> Optimiser) -> MIR -> ISel -> RA -> Emit.

LensPlaysGames replied:
There are two parallel uses of the MIR right now. The whole `MIRValuex86_64` thing is what's used  by `select_instructions` and actually provides enough data to emit machine code from (or at least did). `select_instructions2` is the "dev branch" basically and is where I am actually working on things still. Both of them are used in parallel in different ways by the x86_64 backend right now, and eventually we'll stop using the first one and rename `2` to just `select_instructions`, but baby steps, ya know :P

Sirraide replied:
Ah yeah, I see how it is

LensPlaysGames committed:
[Minor] Fix typo in assert message

LensPlaysGames committed:
[Minor] Simplify implementation of `DBGASSERT` macro

LensPlaysGames committed:
[MIR] Properly inline register and immediate operands when lowering, among other things

Also in this commit: static and local references get their own type of
operand. Not sure if this is strictly necessary, but it solves a bug
regarding having to lookup MIR instruction by vreg, and that's not
something we want to have to do.

LensPlaysGames committed:
Fix lots of implicit fallthrough warnings under GCC

LensPlaysGames committed:
[MIR] Blocks now keep vectors of instructions...

LensPlaysGames committed:
[Minor] Adjust struct layout to save on padding

LensPlaysGames committed:
[MIR/x86_64] `AND`, `OR`, `NOT`

LensPlaysGames committed:
[MIR/ISel/x86_64] ¡`MUL`!

LensPlaysGames committed:
[ISel/x86_64] Return with operand generates move to RAX

LensPlaysGames committed:
[MIR] Move phi2copy from IR to MIR, just after IR->MIR lowering phase

`phi2copy` is now implemented at the general MIR level, meaning RA won't
have to worry about PHIs or any of their nonsense, and neither will
backends when doing instruction seleciton or code emission.

Also, MIR blocks and functions get some helper functions.

LensPlaysGames committed:
[x86_64/ISel] Fix invalid opcode assignment

Also make some fallthroughs explicit

LensPlaysGames committed:
[x86_64/Minor] x86_64 MIROpcode -> string conversion

LensPlaysGames committed:
[MIR] Debug printing handled in more generic way

This allows for backends to extend the printouts with their own opcode
translations, meaning x86_64 op1080 actually becomes `mov` in the
output, which is **much** more readable, to say the least.

LensPlaysGames committed:
[x86_64/ISel] `SUB`

LensPlaysGames committed:
[x86_64/ISel] Comparisons (`SETCC`)

LensPlaysGames committed:
[MIR/Minor] Differentiate name, block, and function operands when printing

LensPlaysGames committed:
[x86_64/ISel] Use block operands instead of name for conditional jumps

Slowly we are relying less and less on the "origin" IR pointers for
actual functionality. Poggers.

BTW we can always revert this, since the machine instructions don't
care whether it's a block or not and that might needlessly complicate
final code emission. Either way we could now do that by taking
advantage of the actual block operands instead of subverting them.

LensPlaysGames committed:
[x86_64/ISel] Regular branches now retain block operand vs name

This may be changed back, much like the conditional branch change, but
we want to retain as much information as possible for going into RA,
right now.

LensPlaysGames committed:
[MIR] Call arguments

LensPlaysGames committed:
[MIR] Fix bug regarding MIROperand SSO

We were restoring the old operands but not adding the new one, when
going from small to large.

LensPlaysGames committed:
[Utils] `FUNCTION_POINTER` macro for easier typedefs

I just can't for the life of me really mentally parse the function
pointer typedef syntax, and I always forget how to do it right. This
way there's no confusion :)

LensPlaysGames committed:
[MIR] ¡Frame Objects!

LensPlaysGames committed:
[MIR] The rest of the owl

I forgot this code in the last commit, haha.

Also fix x86_64/ISel bug with `lea` not using the new stack allocation
setup.

LensPlaysGames committed:
[x86_64/ISel] Fix introduced bug to alloca (lol)

Patch one hole, another opens up...

LensPlaysGames committed:
I am mentally burned and will not be caring about this commit message

LensPlaysGames committed:
[RA] Begin the overhaul to use MIR instead of IR :sob:

LensPlaysGames committed:
[RA] Cries from a soul in wretched pain

LensPlaysGames committed:
[MIR] Retain Control Flow Graph through to lowered MIR

This means keeping track of each MIRBlock's predecessors, mainly, as
well as adding a `lowered` field so that references to blocks in the
general MIR may be translated into blocks in the lowered MIR.

LensPlaysGames committed:
[RA] Populate list of vregs

LensPlaysGames committed:
[Utils] Rearrange includes to separate std and ours

LensPlaysGames committed:
[Error] Move `seek_location` from parser to error files

Not sure why it was in the parser in the first place... either way,
this allows `error.c` to no longer rely on parser.h, which is a good
dependency to break if you ask me (especially when attempting to reuse
this code for other projects)

LensPlaysGames committed:
[x86_64] Fallback to printing common opcode mnemonic when not found

LensPlaysGames committed:
[MIR] MIR register operand gets `defining_use` boolean member

This will allow register allocation to actually keep track of the
defining use of each virtual register, and therefore properly
calculate live ranges.

LensPlaysGames committed:
[RA] Migration towards MIR (it's actually working POG)

LensPlaysGames committed:
[Minor/RA] Comments, `DEBUG` instead of `print`

LensPlaysGames committed:
[x86_64/ISel] Virtual register defining use for comparison

LensPlaysGames committed:
[RA] Retain virtual register size information

LensPlaysGames committed:
[MIR] Small bugfixes

LensPlaysGames committed:
[x86_64/GObj] `add` in `imm to reg` form and `ret`

:^&

Also fix a bug regarding accidental fallthrough

LensPlaysGames committed:
Minor Bugfixes

RA had an issue with invalid format specifiers since the vreg size
inclusion refactor.

Some fallthrough attributes weren't followed by a semi-colon, and
that's a bad thing ig.

Size of source register is properly set when instruction-selecting a
store instruction (when storing into a static ref). I'm still not sure
how we are going to handle type-based size information in the ISel DSL.

LensPlaysGames committed:
[Docs/ISel] A very rudimentary thought-dump on ISel

LensPlaysGames committed:
Fix some warnings

LensPlaysGames committed:
[x86_64/GObj] Choose smaller add opcode when immediate is within range

This shaves three bytes off of every single immediate add/subtract,
pretty much.

LensPlaysGames committed:
[GObj] `delete` function to free heap-allocated resources

Less memory leaks is always a good time :^&

LensPlaysGames committed:
[MIR] Basic operand count and operand kind matching function

LensPlaysGames committed:
[Examples] Fix old-style comment starter in `Hello, world!`

A while ago we changed the line comment starter to `;;` and `;` by
itself became an expression delimiter.

LensPlaysGames committed:
[x86_64] Bugfix: vector_insert_after takes pointer iterator, not index

LensPlaysGames committed:
[x86_64] Operand matching makes for 1000x better emission code

LensPlaysGames committed:
[x86_64] Replace switch-hell with basic matching goodness

LensPlaysGames committed:
[x86_64] Fix major accidental fallthrough bug

Also fix a copy-and-paste error where `femit` was used instead of
`mcode` for the function prefix.

LensPlaysGames committed:
[MIR] PHIs require immediates to be emitted as instructions

We remove them after lowering out PHIs from the MIR, but to do that
lowering we require immediates to have a corresponding machine
instruction. The same issue may arise if an IR_REGISTER is the
argument of a phi node, so we may want to *always* generate at least
one general MIR instruction from each and every IR instruction.

LensPlaysGames committed:
[x86_64/ISel] Remove RBP as an operand for locals, among other things

A local will /always/ be offset from RBP so we don't need to encode
that in the lowered MIR.

There's also some hacky warnings in place because a lot of registers
have a size of zero for some reason, and I'm trying to figure out the
cause of that.

Also in this commit is `local to reg` moves for both mcode and assembly
targets on x86_64 architecture.

LensPlaysGames committed:
[RA] When removing a live value, ignore register size, just compare value

This requires a custom remove_element_unordered function but that's no
problem!

Also, remove some code that hasn't been used in a /while/; I don't
think we'll need an IRInstruction list any time soon.

LensPlaysGames committed:
[ISel] Make a big, huge mess that likely doesn't work yet :^&

LensPlaysGames committed:
[ISel] Fix bug regarding invalid case in switch

TOKEN_SEMICOLON is just an enum, not equal to an actual semicolon.

LensPlaysGames committed:
[ISel] Progress is progress

LensPlaysGames committed:
[ISel] `consume`, `ERR`, parsing

LensPlaysGames committed:
[MIR] Printing MIR is hard, mmkay

I don't technically think we *need* this but it helps to be able to
print instructions/operands not connected to a function. We can
probably get rid of this abstraction at any point, but it's here just
for devving ISel right now

LensPlaysGames committed:
[ISel] A basic string hash table to store environment things

LensPlaysGames committed:
`WARN` macro

LensPlaysGames committed:
[ISel] Opcodes are getting resolved!

LensPlaysGames committed:
[ISel] Some parsing changes

LensPlaysGames committed:
[Minor] Formatting, small bugfix

`%s` on a non NULL-terminated string is a bad idea!

Don't print debug output when adding new opcodes/opkinds to isel
environment.

LensPlaysGames committed:
[Codegen] More ISel progress

We are now able to handle arch-specific environment initialisation as
well as utilise it with x86_64 to properly parse `MX64` opcodes.

LensPlaysGames committed:
[ISel] Huge amounts of progress

Okay, with this iteration, `isel_do_selection` should actually
function ... :eyes:

LensPlaysGames committed:
[x86_64] Utilise ISel, begin to cleanup

Get rid of old isel ways: implementing patterns in code
absolutely *sucks*, and I'm glad to never have to do it again.

I still want to remove the `select_instructions` behemoth, but I want
to retain that to build isel patterns off of until the ISel is up to
snuff with what it was before.

LensPlaysGames committed:
[ISel] Even more progress

We've added register-type values to the isel environments, that way
backends can reference registers directly if need be.

We still need some way to actually specificy for each MIROperand
whether or not we should be comparing the value to the pattern's
value.

Fixed broken sorting algorithm (use `j` instead of `i` hehe).

LensPlaysGames committed:
[ISel] Less debug printouts

LensPlaysGames committed:
[ISel] Mark defining uses

LensPlaysGames committed:
[ISel] Fix bug regarding vreg numbering and RA

LensPlaysGames committed:
Lots of small fixes

LensPlaysGames committed:
[RA] Use `FOREACH_MIR_OPERAND`instead of manual loops

LensPlaysGames committed:
[MIR] Fix clashes between RA `VReg` and `MIROperandRegister`

This isn't necessarily ideal---it would be much better if
MIROperandRegister fit inside 8 bytes---however the fact that our
`vector_contains` and such comparison functions don't have that strong
of type-checking, this is what will be done (for now) to make sure we
don't shoot ourselves in the foot with that one in RA (as I have
  already done somewhere, evidently).

LensPlaysGames committed:
[x86_64] Print MIR after RA

LensPlaysGames committed:
[MIR] Fixup references to pre-coloured IR instructions

LensPlaysGames committed:
[MIR] Use pre-coloured register as instruction register if possible

LensPlaysGames committed:
[x86_64/IR] Set return register

LensPlaysGames committed:
[x86_64] Function epilogue/prologue functionality

LensPlaysGames committed:
[x86_64] Callee-saved register restoration

This isn't *quite* in effect due to RA not having `track_registers` up
to snuff yet, but we're making progress.

Also, we now never emit `sub $0, %rsp` at beginning of function.

LensPlaysGames committed:
[x86_64/ISel] ¡More patterns!

LensPlaysGames committed:
[MIR] Some helpers

LensPlaysGames committed:
[x86_64] Remove register-to-register moves of equal value+size

LensPlaysGames committed:
[MIR] Fix missing field initiliasers warning

LensPlaysGames committed:
[RA] Fix unused variable warning due to sometimes-disabled debug code

LensPlaysGames committed:
[ISel] Fixup parameter name to no longer conflict with global symbol

LensPlaysGames committed:
[x86_64] Separate targets from main arch file

LensPlaysGames committed:
[x86_64] Remove `enum Instruction` in favor of `MIROpcodex86_64`

LensPlaysGames committed:
[ISel] Fix bug regarding use of non NULL-terminated string

LensPlaysGames committed:
[x86_64/ISel] More patterns, more code emission

LensPlaysGames committed:
[CMake] Set working directory of tests

LensPlaysGames committed:
[x86_64/GObj] Assemble shift instructions

LensPlaysGames committed:
[x86_64/ISel] ¡Ever-yet more patterns!

LensPlaysGames committed:
Handle `extern` functions

LensPlaysGames committed:
[RA] Reenable register tracking

LensPlaysGames committed:
[x86_64] Get rid of extra loop by merging two alike ones

We can mangle function names and do basic block nonsense all at once!

LensPlaysGames committed:
[x86_64] <=¡Calls!=>

LensPlaysGames committed:
[x86_64] More patterns, more code emission

LensPlaysGames committed:
[Codegen] Get rid of *very* old register allocation code

Oh, the nostalgia of when RA didn't haunt me in my nightmares

LensPlaysGames committed:
[CMake] Temporarily remove `codegen/ir.c` from build

LensPlaysGames committed:
[Codegen] Refactor target specification

I'm still not 100% happy with the system; ideally, a user would be
able to specify all three in one go, like with gcc/llvm target
triples.

NOTE: This very well may not compile; I've changed a lot of things and
may not have been perfect in unweaving the different changes.

LensPlaysGames committed:
[x86_64] Insert copies to argument registers

LensPlaysGames committed:
[x86_64] Fix some bugs regarding final lowering of calls

LensPlaysGames committed:
[x86_64] Actually calculate stack offsets

LensPlaysGames committed:
[x86_64] Code emission of statics

LensPlaysGames committed:
[x86_64/ISel] More memory-related patterns

LensPlaysGames committed:
[Minor/x86_64] Fix invalid format specifiers

LensPlaysGames committed:
[Minor/ISel] Reorganise some code

LensPlaysGames committed:
[ISel] Allow emitted instructions to reference themselves and previous

LensPlaysGames committed:
[x86_64] Code emission fixups regarding statics

LensPlaysGames committed:
[x86_64] Get rid of weird `calls_to_fixup` indirection

LensPlaysGames committed:
[GObj] Generic object API for `FILE*`, not just file path

LensPlaysGames committed:
[Utils] I'll take "things you can't believe aren't in the c standard library" for $200, Alex

LensPlaysGames committed:
Automatically generate output filepath from input filepath

LensPlaysGames committed:
[x86_64/Bugfix] Fix improperly renamed variable

We renamed the inner-most variable `i` to `r`, but apparently I also
accidentally renamed the non-inner-most variable `i` to `r`. This is
why we avoid shadowing!!

LensPlaysGames committed:
[x86_64/Bugfix] Fix borked if condition that caused stack misalignment

LensPlaysGames committed:
[x86_64/ISel] Whoops, we get a pointer with `lea`, not `mov`

LensPlaysGames committed:
Add missing command line argument parsing

It was in the help, but not actually being checked for.

LensPlaysGames committed:
[x86_64/ISel] Bugfix: Mismatched comparison type

LensPlaysGames committed:
[x86_64/ISel] Simple pattern for copy of immediate

LensPlaysGames committed:
[MIR] Get rid of weird checks for copies

LensPlaysGames committed:
[x86_64] No longer write `.data` assembly to object files (lol)

LensPlaysGames committed:
[x86_64] Merge stack offset calculation loop with post-RA fixup

LensPlaysGames committed:
[x86_64/Bugfix] Remove code that should *not* have been there

O copy and paste, for better and worse

LensPlaysGames committed:
[RA/Bugfix] All vreg operands of an instruction interfere with each other

LensPlaysGames committed:
[Codegen/Bugfix] Fix major bug due to a single missing byte

Opening the output code file in binary mode is *critical* to the
proper functionality of backends (on Windows).

LensPlaysGames committed:
[x86_64/GNUAsm] `AND`, `JMP`, `CMP`, `TEST`, `SETCC`, `JCC`

LensPlaysGames committed:
[x86_64/GObj] Assemble `AND` in `reg to reg` form

LensPlaysGames committed:
[x86_64/GObj] `AND`, `JMP`, `CMP`, `TEST`, `SETCC`, `JCC`

LensPlaysGames committed:
[x86_64, GObj] ELF x86_64 uses PLT32 relocation for functions

ffs... I don't understand why something called `R_X86_64_PLT32` has
seemingly /nothing/ to do with the PLT on x86_64. Like, x86_64 is
already in the enum name ... you couldn't have called it
`R_X86_64_BRANCH32`??

Quote: H.J. Lu from Binutils Git commit bd7ab16b4537788ad53521c45469a1bdae84ad4a message
<https://sourceware.org/git/?p=binutils-gdb.git;a=commitdiff;h=bd7ab16b4537788ad53521c45469a1bdae84ad4a;hp=80c96350467f23a54546580b3e2b67a65ec65b66>
> Since there is no need to prepare for PLT branch on x86-64, generate
> R_X86_64_PLT32, instead of R_X86_64_PC32, if possible, which can be
> used as a marker for 32-bit PC-relative branches.
>
> [...]
>
> On i386, there are 2 types of PLTs, PIC and non-PIC.  PIE and shared
> objects must use PIC PLT.  To use PIC PLT, you need to load
> _GLOBAL_OFFSET_TABLE_ into EBX first.  There is no need for that on
> x86-64 since x86-64 uses PC-relative PLT.
>
> On x86-64, for 32-bit PC-relative branches, we can generate PLT32
> relocation, instead of PC32 relocation, which can also be used as
> a marker for 32-bit PC-relative branches.  Linker can always reduce
> PLT32 relocation to PC32 if function is defined locally.   Local
> functions should use PC32 relocation.  As far as Linux kernel is
> concerned, R_X86_64_PLT32 can be treated the same as R_X86_64_PC32
> since Linux kernel doesn't use PLT.

I don't really know what this means, but basically what I've pieced
together from a few different sources is that on x86_64 PLT32
relocations are used to mark PC relative **branches**, not just a PC
relative relocation (which it is). The linker implements it
effectively the same as PC32 relocation except for a very niche corner
case (afaik, again I'm not an expert).

LensPlaysGames committed:
[x86_64] Stack frame is now handled by emission, not in MIR at all

This honestly cleans up the code the most, imo

LensPlaysGames committed:
[Vector] Remove `vector_append_all`, favor `vector_append`

Sirraide:
> Btw, speaking of macros that should be deleted: `vector_append`, NOT
> `vector_append_all`, should also be yeeted because they do the same
> thing, the latter is actually used, and the former uses `->` instead
> of `.`.

LensPlaysGames committed:
[Vector] Fix possible UAF in `vector_insert`

Because `vector_reserve` may actually re-allocate, using a pointer to
an element from before the call boundary is dangerous/UB. This alters
insert to get the index before calling reserve, then recalculate the
pointer after using it.

Sirraide generously offered the following:
```c
  size_t _index = (pos) - (vector).data;
  vector_reserve((vector), 1);
  memmove(
      (vector).data + _index + 1,
      (vector).data + _index,
      ((vector).size - _index) * sizeof *(vector).data
  );
  (vector).data[_index] = element;
  (vector).size++;
```

LensPlaysGames committed:
[x86_64/GObj] Scaffolding for `IMUL` in `reg to reg` form

LensPlaysGames committed:
[Tests] `local_integer.int`

I can't believe I hadn't committed this yet; it's had to have been at
least a month since I've written this test, lol.

LensPlaysGames committed:
[Docs] Compilation overview

LensPlaysGames committed:
[Docs/CompilationOverview] `tikz` is hard, okay?

LensPlaysGames committed:
[Docs/CompilationOverview] First pass at improvement, add asm example

Also fix figures just kind of floating into the next section because
they feel like it.

LensPlaysGames committed:
[x86_64/GObj] Fix typo (was using wrong section size to get byte offset of symbol)

LensPlaysGames committed:
[GObj] Print function

LensPlaysGames committed:
[Docs/CompilationOverview] Distinguish frontend/backend

LensPlaysGames committed:
[x86_64] Fix bug(s) regarding locals

Thanks KingBunz :^)

LensPlaysGames committed:
Rename `--debug-ir` to `--print-ir` to be more consistent

LensPlaysGames committed:
[x86_64] Get rid of patterns generating dead code

We *really* need that `discard` feature, haha

LensPlaysGames committed:
[Docs/CompilationOverview] Add IR example

LensPlaysGames committed:
[ISel] Add `discard` keyword to emit no instructions for a given pattern

LensPlaysGames committed:
[x86_64/ISel] Take advantage of new `discard` keyword

Remove old no-ops when we had to just ignore instructions that should
be discarded during ISel.

Also fix a bug with the MIR_LOAD(Register) pattern, as we need to have
a "mem to reg" form, which means (Register address, Immediate offset,
Register destination) operands.

LensPlaysGames committed:
[Minor/RA] Use `FOREACH_MIR_OPERAND` macro instead of doing it manually

LensPlaysGames committed:
[Docs/CompilationOverview] Add `\noindent` everywhere, lest Sirraide tear my head off

LensPlaysGames committed:
[TODO/DONE] `LINUX` calling convention renamed to `SYSV`

LensPlaysGames committed:
[x86_64/ISel] Get pointer of frame object, dereferencing store

LensPlaysGames committed:
[x86_64/ISel/Bugfix] Operands of `cmp` were backwards, lol

LensPlaysGames committed:
[RA] (Hopefully) a bugfix regarding defining uses and interference

`MOV v1, v2 DEF`

As you can see, it's quite clear that if we move v1 into v2, they can
actually be the same register. So defining uses of register operands
shouldn't interfere with any other virtual register operands (by
default). This may be catastrophically wrong, but it does fix some RA
bugs in `simple_math.int` where it was assigning interference where
there shouldn't be any.

LensPlaysGames committed:
[x86_64] `CQO`

LensPlaysGames committed:
[ISel + RA] MIRInstructions gain `clobbers`

LensPlaysGames committed:
[IR] Don't emit a million static variables at the beginning of every function

Register allocation and use of vregs has differed, now. Any reference
to a static as an operand gets inlined during IR->gMIR lowering, and
therefore we don't actually need these here to load the pointer into a
register, as we did before.

LensPlaysGames committed:
[RA] Rearrange interference setting as it was slightly wrong

LensPlaysGames committed:
[x86_64] Saving/restoration of callee-saved registers

LensPlaysGames committed:
[x86_64/ISel] Mention bug, add `div` and `mod` patterns

LensPlaysGames committed:
[x86_64/Asm] `imm_to_mem` requires a size

We still need to actually use it to do the `QWORD PTR`/`q` suffix
things in the emitted assembly, but we're one step closer to actually
being able to do that.

LensPlaysGames committed:
[MIR] Print clobbers

LensPlaysGames committed:
[MIR + x86_64] When necessary, include size of memory operation

An immediate has no "size", so when storing one into memory we don't
actually know how many bytes to store into memory. This fixes that by
passing the size explicitly. We also now take advantage of this passed
size in the assembly backend.

LensPlaysGames committed:
[x86_64/GObj] Handle new sized memory operand things

LensPlaysGames committed:
[x86_64] More checks for zero-sized registers...

LensPlaysGames committed:
[MIR] More control flow tracking

LensPlaysGames committed:
[ISel] *Actually* follow control flow to mark defining uses of vregs

LensPlaysGames committed:
[RA/Minor] Use `is_entry` instead of "is first in function"

LensPlaysGames committed:
[x86_64/Minor] Fix bug regarding debug mnemonic printout for MIR opcodes

MX64_COUNT is just the count, and there happens to be a general
MIR opcode with that same value (MIR_TRUNCATE, currently). So this
fixes that little bug

LensPlaysGames committed:
¡New Attribute: `discardable`!

LensPlaysGames committed:
[x86_64] Add TODO handling for unhandled MX64 opcodes

LensPlaysGames committed:
[x86_64] Simple `imul` and `idiv`/`div` emission

LensPlaysGames committed:
[Minor/Codegen] Fix Indentation

LensPlaysGames committed:
[MIR] *Actually* handle function references

LensPlaysGames committed:
[x86_64] Function references

LensPlaysGames committed:
[MIR] Better handling of casts

LensPlaysGames committed:
[Editor/Emacs] Update Intercept syntax highlighting

LensPlaysGames committed:
[x86_64/Asm] Fix horrible bug regarding missing `break` keywords

LensPlaysGames committed:
[x86_64/ISel] Pointer arithmetic regarding statics

LensPlaysGames committed:
[x86_64] `SUB` in `imm_to_mem` form

LensPlaysGames committed:
[x86_64] Fix stack-based parameters + ISel pattern to optimise local ptr calculation

Not tested the SysV ABI yet, but MSWIN is working now.

LensPlaysGames committed:
[x86_64/Asm] Don't emit offset when it's zero

LensPlaysGames committed:
[x86_64] Zero out registers, when necessary

LensPlaysGames committed:
[x86_64/ISel] Pattern to make `function_pointer.int` test pass

LensPlaysGames committed:
[x86_64] Error on invalid `mov` operands in reg to reg form

LensPlaysGames committed:
[Tests] `div.int` and `if_expression.int`

LensPlaysGames committed:
[x86_64] For reg to reg moves, check for zero-sized registers

We really need to find a solution to this...

LensPlaysGames committed:
[RA] Include hardware registers in interference graph

LensPlaysGames committed:
[x86_64/GObj] `not` machine code

LensPlaysGames committed:
[x86_64/GObj] `and` machine code in `imm to reg` form

LensPlaysGames committed:
[x86_64/GObj] Zero register for `reg to reg` moves, when necessary

Also implement `not` of a register very simply

LensPlaysGames committed:
[MIR + Utils] `%V` format specifier for MIR V/HW register

LensPlaysGames committed:
[x86_64] Slightly confusing but now working lowering of calls

So instead of copying *into* the result register from the actual
virtual register of the call, we are *replacing* the virtual register
of the call with a hardware register (according to calling convention,
ideally). Then, any uses of the call's **virutal** register are
replaced with a new virtual register; one that refers to a copy of the
result register...

LensPlaysGames committed:
[RA] Proper handling of hardware register interferences

Or at least, closer to proper

LensPlaysGames committed:
[x86_64/ISel] Lots of comparison patterns...

LensPlaysGames committed:
[x86_64] Print less unless prompted

LensPlaysGames committed:
[RA] Don't set all hardware registers used as operands as interfering

Basically, we don't know for sure *when* hardware registers are
defined, so we can't properly set interference excluding
definition (unless we somehow kept track of where hardware registers
get their values set, but that would mean each backend would have to
tell us that, because we don't know opcodes at time of RA, after ISel).

LensPlaysGames committed:
[x86_64/Minor] Add TODO for unhandled SysV ABI case

This comes to light through `tst/tests/function_arguments_many.int`

LensPlaysGames committed:
[x86_64] No longer insert an unused instruction

LensPlaysGames committed:
[Sema] Proper implicit casts for binary arithmetic operators

Integer promotion go brr

LensPlaysGames committed:
[Codegen] Actually use `type_is_signed`, don't just guess

LensPlaysGames committed:
[x86_64] Remove `movzx(r32, r64)` no-op

Sirraide:
> I guess, as an invariant, we can just require in the x86_64 backend that any operation that operates on 32-bit registers clears the upper bits
> If an instruction for some reason doesn’t do that, we can emit an and or sth

Basically, there is one/two undocumented instruction(s) that *may* not
zero the top bits when moving into a 32 bit register, but that's
... *rare*. And also not in use by this compiler. So this seems like a
good invariant to have for the x86_64 backend.

LensPlaysGames committed:
[x86_64/ISel] `not` in `imm` form

LensPlaysGames committed:
[x86_64/Asm] `cwd`, `cdq`, `movsx`

LensPlaysGames committed:
[x86_64/GObj] `cwd`, `cdq`, `cmp`, `movsx`

LensPlaysGames committed:
[x86_64/GObj] `idiv` in `reg` form

LensPlaysGames committed:
[x86_64/GObj] Fix invalid `break` (supposed to be `FALLTHROUGH`)

Oops!

LensPlaysGames committed:
[x86_64/SysV] Error upon failure to lower a register class

LensPlaysGames committed:
[Tests] Simplify now that we have array initialisation

LensPlaysGames committed:
[x86_64/GObj] `movzx` machine code

LensPlaysGames committed:
[x86_64/GObj] BugFix: Add missing `break`

LensPlaysGames committed:
[x86_64/ISel] Branch conditional with immediate condition operand

Kind of ridiculous, but for debugging purposes...

LensPlaysGames committed:
[Tests] Update ALGOL 68 test suite

LensPlaysGames committed:
[Tests] Update C++ test suite to allow different intc targets

This means we can now test both object file and assembly backends
using ctest, as well as optimised versions.

LensPlaysGames committed:
[Tests] Remove object file testing from default suite

Testing the same test in multiple backends is *not* something we need
to be doing every time we make a small change; it should be done, but
not all the time during development, ig.

LensPlaysGames committed:
[Tests] Fix concurrency issue with file names in C++ test driver

LensPlaysGames committed:
[Tests] Update `lots-of-phi-nodes.int` to work with new language semantics

LensPlaysGames committed:
[x86_64/GObj] Zero-sized registers will be the bane of my existence

I should really just create a macro that does this and apply it to
every register operand, lolol.

LensPlaysGames committed:
[x86_64/ISel] `sub` in `reg to reg` form

LensPlaysGames committed:
[Codegen] `argv` is now `@@byte`, not `@@integer`

LensPlaysGames committed:
[Tests] Delete (possibly) created file upon failure
